### PR TITLE
Storage object deletion waits for consistency.

### LINKED ISSUE #344: Storage delete operations are not sync
As pointed out by @craigcitro, `Object.delete` and `Bucket.delete` actually submit delete jobs to the storage service, so they're not exactly synchronous operations, which is different from user expectations.

We should explore either waiting on the jobs to finish as part of the API itself, or rename these operations to add a `_async` suffix.

### PR DESCRIPTION
**Pull request by craigcitro**
Fixes #344.

GCS object deletions are eventually consistent, in that deleting an object in
a globally-available bucket may result in the object still appearing when
listing objects in the bucket:
  https://cloud.google.com/storage/docs/consistency#eventually_consistent_operations

This is a detail, but can lead to confusion for users: someone following the
examples in the tutorial notebooks may see rare-but-regular failures about an
object they just deleted continuing to exist. (We've seen this running the
notebooks as automated tests.) This PR teaches an object to poll the bucket
until it no longer appears. This is controllable by `wait_for_deletion`, and
defaults to **on**.

I've added two tests: first is a mock-based unit test, which works; the second
is intended as a first step towards integration tests. In this case, we assume
that the user running the test has ambient credentials that `google.datalab`
will discover.

### CODE REVIEW

**Comment by craigcitro** [comment]
PTAL @yebrahim

File: google/datalab/storage/_object.py [hunk 1]
```diff
@@ -147,6 +156,14 @@ def delete(self):
         self._api.objects_delete(self._bucket, self._key)
       except Exception as e:
         raise e
+      if wait_for_deletion:
+        while True:
+          objects = Objects(self._bucket, prefix=self.key, delimiter='/',
+                            context=self._context)
+          if any(o.key == self.key for o in objects):
+            time.sleep(_POLLING_SLEEP)
+            continue
```
Comment (yebrahim) [inline google/datalab/storage/_object.py]: Do we want to have a timeout for this or are we sure the API never returns except in the success case?

File: tests/integration/storage_test.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    if not project_id:
+      # If there's no project ID set, we try asking gcloud. We need to send
+      # stderr to /dev/null, otherwise we get some spurious output.
+      gcloud_output = subprocess.check_output([
+          'gcloud', 'config', 'get-value', 'project'], stderr=subprocess.PIPE)
+      project_id = gcloud_output.strip()
+    logging.info('Using project ID: %s', project_id)
+
+    context = google.datalab.Context.default()
+    context.set_project_id(project_id)
```
Comment (yebrahim) [inline tests/integration/storage_test.py]: Why don't we just use `google.datalab.Context.default()` as is? It should pick up the project value from gcloud, just like what the lines 21-31 do.

Comment (yebrahim) [inline tests/integration/storage_test.py]: We should separately have a unit test that makes sure setting a project on gcloud gets read by the `Context.default()` object, so using it directly here should be fine.

File: google/datalab/storage/_object.py (hunk 1 above)
Comment (craigcitro) [inline google/datalab/storage/_object.py]: Ah, good call. Done.

File: tests/integration/storage_test.py (hunk 2 above)
Comment (craigcitro) [inline tests/integration/storage_test.py]: Oh, nice! I didn't realize that at all.

Fixed.

**Comment by craigcitro** [comment]
I'm happy to add polling on create operations -- my feeling was that we should let this bake, and if it seems to work well, actually add it for *all* eventually consistent operations in one go. Should I file a separate bug to that effect?

File: tests/integration/storage_test.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+class StorageTest(unittest.TestCase):
+
+  def setUp(self):
+    context = google.datalab.Context.default()
+    logging.info('Using project: %s', context.project_id)
+    self._context = context
+
+    suffix = ''.join(random.choice(string.letters) for _ in range(8))
+    self._test_bucket_name = '-'.join((project_id, suffix)).lower()
```
Comment (yebrahim) [inline tests/integration/storage_test.py]: `context.project_id`

**Comment by yebrahim** [comment]
Sweet! Yes I think a follow-up issue on eventual consistency should be enough for now.

**Comment by craigcitro** [comment]
https://github.com/googledatalab/pydatalab/issues/355 is the follow-on.

File: tests/integration/storage_test.py (hunk 3 above)
Comment (craigcitro) [inline tests/integration/storage_test.py]: Done.

**Pull request merged**